File: solutions/2026-05-06/cpp/2026-05-06-rotating-the-box.cpp
```cpp
// Summary: Rotates a 2D grid representing a box 90 degrees clockwise and simulates gravity for stones.
// Link: https://leetcode.com/problems/rotating-the-box/
// Approach:
// 1. Transpose the grid: Convert the m x n grid to an n x m grid by swapping rows and columns.
//    This effectively simulates the initial 90-degree clockwise rotation.
// 2. Simulate Gravity for each column of the transposed grid:
//    For each column, iterate from the bottom up. Maintain a 'free_row' pointer that indicates
//    the next available empty spot from the bottom.
//    - If a stone ('#') is encountered, place it at 'free_row' and decrement 'free_row'.
//    - If an obstacle ('*') is encountered, update 'free_row' to be one position above the obstacle.
//    - If an empty spot ('.') is encountered, it's skipped, and 'free_row' remains unchanged.
// 3. The resulting grid after gravity simulation is the final rotated and settled box.
// Time Complexity: O(m * n) - We iterate through the grid multiple times (transpose and gravity simulation),
//   but each cell is visited a constant number of times.
// Space Complexity: O(m * n) - To store the transposed and rotated grid.
#include <vector>
#include <string>
#include <algorithm>

class Solution {
public:
    std::vector<std::vector<char>> rotateTheBox(std::vector<std::vector<char>>& boxGrid) {
        int m = boxGrid.size();
        int n = boxGrid[0].size();

        // Step 1: Transpose the grid to simulate 90-degree clockwise rotation
        // The new grid will have dimensions n x m
        std::vector<std::vector<char>> rotatedGrid(n, std::vector<char>(m));
        for (int r = 0; r < m; ++r) {
            for (int c = 0; c < n; ++c) {
                // After 90-degree clockwise rotation, original (r, c) goes to (c, m-1-r)
                rotatedGrid[c][m - 1 - r] = boxGrid[r][c];
            }
        }

        // Update dimensions for the rotated grid
        int new_m = n; // number of rows in rotatedGrid
        int new_n = m; // number of columns in rotatedGrid

        // Step 2: Simulate gravity for each column of the rotated grid
        for (int c = 0; c < new_n; ++c) {
            // 'free_row' keeps track of the lowest available empty spot in the current column
            // Initialize it to the bottom of the column
            int free_row = new_m - 1;

            // Iterate from the bottom up for the current column
            for (int r = new_m - 1; r >= 0; --r) {
                if (rotatedGrid[r][c] == '#') {
                    // If we find a stone, move it to the 'free_row'
                    // and update 'free_row' to the next available spot above it.
                    // We only swap if the stone is not already at the free_row to avoid unnecessary operations.
                    if (r != free_row) {
                        rotatedGrid[free_row][c] = '#';
                        rotatedGrid[r][c] = '.';
                    }
                    free_row--; // Move the free spot upwards
                } else if (rotatedGrid[r][c] == '*') {
                    // If we encounter an obstacle, it stops stones from falling further.
                    // The next available free spot will be just above this obstacle.
                    free_row = r - 1;
                }
                // If rotatedGrid[r][c] is '.', it's already an empty spot,
                // and 'free_row' doesn't need to be updated based on it,
                // unless it's the first empty spot encountered from the bottom.
            }
        }

        return rotatedGrid;
    }
};
```

File: solutions/2026-05-06/javascript/2026-05-06-rotating-the-box.js
/**
 * @summary Rotates a box grid 90 degrees clockwise and simulates gravity for stones.
 * @link https://leetcode.com/problems/rotating-the-box/
 *
 * Approach:
 * 1. Transpose the grid: The original grid `boxGrid` (m x n) becomes `transposedGrid` (n x m).
 *    The cell `boxGrid[r][c]` moves to `transposedGrid[c][r]`.
 * 2. Reverse rows of the transposed grid: To simulate the 90-degree clockwise rotation,
 *    we reverse each row of the `transposedGrid`. Now `transposedGrid[c][r]` becomes
 *    `rotatedGrid[c][n-1-r]`. This effectively places the elements in their correct
 *    rotated positions.
 * 3. Simulate gravity: For each column `j` in the `rotatedGrid`, we iterate from bottom
 *    to top. We maintain a pointer `stoneRow` that indicates the next available row
 *    from the bottom where a stone can land.
 *    - If we encounter a stone ('#'), we place it at `rotatedGrid[stoneRow][j]` and
 *      decrement `stoneRow`.
 *    - If we encounter an obstacle ('*'), we update `stoneRow` to be the row just above
 *      the obstacle.
 *    - Empty cells ('.') are naturally handled as we skip over them and only place stones
 *      in available slots determined by `stoneRow`.
 *
 * Time Complexity:
 * - Transposing: O(m * n)
 * - Reversing rows: O(m * n)
 * - Simulating gravity: O(m * n) (each cell is visited a constant number of times)
 * Total Time Complexity: O(m * n)
 *
 * Space Complexity:
 * - Storing the rotated grid: O(m * n) to create the new `rotatedGrid`.
 * Total Space Complexity: O(m * n)
 */

/**
 * @param {character[][]} boxGrid
 * @return {character[][]}
 */
var rotateAndSimulateGravity = function(boxGrid) {
    const m = boxGrid.length;
    const n = boxGrid[0].length;

    // 1. Transpose the grid
    // new grid will be n x m
    const transposedGrid = Array(n).fill(0).map(() => Array(m).fill('.'));
    for (let r = 0; r < m; r++) {
        for (let c = 0; c < n; c++) {
            transposedGrid[c][r] = boxGrid[r][c];
        }
    }

    // 2. Reverse rows of the transposed grid to achieve 90-degree clockwise rotation
    // The dimensions are now n x m, and we want to reverse each row.
    const rotatedGrid = Array(n).fill(0).map(() => Array(m).fill('.'));
    for (let r = 0; r < n; r++) {
        for (let c = 0; c < m; c++) {
            rotatedGrid[r][c] = transposedGrid[r][m - 1 - c];
        }
    }

    // 3. Simulate gravity for each column
    // Iterate through each column of the rotated grid
    for (let c = 0; c < m; c++) {
        // `stoneRow` tracks the next available row from the bottom for a stone to land
        let stoneRow = n - 1;
        // Iterate from the bottom row upwards for the current column
        for (let r = n - 1; r >= 0; r--) {
            if (rotatedGrid[r][c] === '#') {
                // If it's a stone, move it to the `stoneRow` position
                // and decrement `stoneRow` to find the next available spot
                rotatedGrid[stoneRow][c] = '#';
                // Clear the original position of the stone if it's not already the target position
                if (stoneRow !== r) {
                    rotatedGrid[r][c] = '.';
                }
                stoneRow--;
            } else if (rotatedGrid[r][c] === '*') {
                // If it's an obstacle, update `stoneRow` to be the row directly above it.
                // This ensures stones will stack on top of obstacles.
                stoneRow = r - 1;
            }
            // If it's an empty cell '.', we just continue, `stoneRow` remains unchanged.
        }
    }

    return rotatedGrid;
};
```